File: src/routing/handlers/chat-member.ts
import type { Update } from "../../kernel/update.js";
import { CallbackContext } from "../../kernel/context.js";
import { BaseHandler, HandlerCallback } from "./base.js";

/**
 * Handler for chat membership changes (`chat_member` and `my_chat_member` updates).
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 * @example
 * ```ts
 * const memberHandler = new ChatMemberHandler(async (update, context) => {
 *   console.log("Chat member updated:", update.chat_member?.new_chat_member.status);
 * }, ChatMemberHandler.CHAT_MEMBER);
 * ```
 */
export class ChatMemberHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /** Target only member updates of other users (`chat_member`) */
  public static readonly CHAT_MEMBER = 1;
  /** Target only updates for the bot itself (`my_chat_member`) */
  public static readonly MY_CHAT_MEMBER = 2;
  /** Target any chat member update */
  public static readonly ANY = 3;
  /**
   * Filter mask specifying which chat member updates to handle.
   */
  public readonly chatMemberTypes: number;

  /**
   * Constructs a new {@link ChatMemberHandler}.
   *
   * @param callback - Function invoked when the update matches.
   * @param chatMemberTypes - Type filter mask (`CHAT_MEMBER`, `MY_CHAT_MEMBER`, or `ANY`).
   * @defaultValue `ChatMemberHandler.ANY`
   */
  constructor(callback: HandlerCallback<C, R>, chatMemberTypes: number = ChatMemberHandler.ANY) {
    super(callback);
    this.chatMemberTypes = chatMemberTypes;
  }

  /**
   * Checks whether the update matches the requested chat member types.
   *
   * @param update - The update to test.
   * @returns `true` if update matches.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    if (this.chatMemberTypes === ChatMemberHandler.CHAT_MEMBER) {
      return Boolean(update.chat_member);
    }

    if (this.chatMemberTypes === ChatMemberHandler.MY_CHAT_MEMBER) {
      return Boolean(update.my_chat_member);
    }

    return Boolean(update.chat_member || update.my_chat_member);
  }
}

/**
 * Handler for user responses to non-anonymous Telegram polls (`poll_answer` updates).
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 * @example
 * ```ts
 * const pollHandler = new PollAnswerHandler(async (update, context) => {
 *   console.log(`User ${update.poll_answer?.user?.id} voted ${update.poll_answer?.option_ids}`);
 * });
 * ```
 */
export class PollAnswerHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /**
   * Checks whether the update contains a `poll_answer` update.
   *
   * @param update - The update to test.
   * @returns `true` if update has `poll_answer`.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    return Boolean(update.poll_answer);
  }
}

File: packages/python/src/telebot_py/routing/handlers/chat_member.py
"""Handlers for chat membership changes and poll answers."""

from __future__ import annotations

from telebot_py.routing.handlers.base import BaseHandler, HandlerCallback
from telebot_py.types import Update


class ChatMemberHandler(BaseHandler):
    """Handler for chat membership changes.

    Matches ``chat_member`` and/or ``my_chat_member`` updates depending on
    the :attr:`chat_member_types` filter mask, mirroring the node
    ``ChatMemberHandler`` constants.

    Attributes:
        chat_member_types: Filter mask selecting which chat member updates
            to handle (:attr:`CHAT_MEMBER`, :attr:`MY_CHAT_MEMBER`, or
            :attr:`ANY`).
    """

    #: Target only member updates of other users (``chat_member``).
    CHAT_MEMBER = 1
    #: Target only updates for the bot itself (``my_chat_member``).
    MY_CHAT_MEMBER = 2
    #: Target any chat member update.
    ANY = 3

    def __init__(
        self,
        callback: HandlerCallback,
        chat_member_types: int = ANY,
        *,
        block: bool = True,
    ) -> None:
        """Initialize the handler.

        Args:
            callback: Callable invoked when a matching member update arrives.
            chat_member_types: Filter mask (`CHAT_MEMBER`, `MY_CHAT_MEMBER`,
                or `ANY`). Defaults to `ChatMemberHandler.ANY`.
            block: Whether dispatch awaits the callback before moving on.
        """
        super().__init__(callback, block=block)
        self.chat_member_types = chat_member_types

    def check_update(self, update: object) -> bool:
        """Return whether the update matches the requested member types.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the update carries a chat member update allowed by
            the configured mask, ``False`` otherwise.
        """
        if not isinstance(update, Update):
            return False
        if self.chat_member_types == ChatMemberHandler.CHAT_MEMBER:
            return update.chat_member is not None
        if self.chat_member_types == ChatMemberHandler.MY_CHAT_MEMBER:
            return update.my_chat_member is not None
        return update.chat_member is not None or update.my_chat_member is not None


class PollAnswerHandler(BaseHandler):
    """Handler for user responses to non-anonymous polls.

    Matches ``poll_answer`` updates, mirroring the node ``PollAnswerHandler``.
    """

    def check_update(self, update: object) -> bool:
        """Return whether the update carries a poll answer.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the update carries a ``poll_answer`` payload.
        """
        return isinstance(update, Update) and update.poll_answer is not None
